usr/src/uts/common/brand/lx/procfs/lx_prvnops.c

```diff
@@ -7660,6 +7775,127 @@ lxpr_write_tcp_property(lxpr_node_t *lxpnp, struct uio *uio,
 	return (res);
 }
 
+/*
+ * lxpr_write_netstacks_property will set attribute @prop for all netstacks
+ * to the supplied value. If the value is out of range, then no update is done.
+ */
+#define	PTBL_FROM_NETSTACK(ns, proto)\
+	((proto) == MOD_PROTO_UDP  ? (ns)->netstack_udp->us_propinfo_tbl :\
+	(proto) == MOD_PROTO_SCTP  ? (ns)->netstack_sctp->sctps_propinfo_tbl :\
+	(proto) == MOD_PROTO_RAWIP ? (ns)->netstack_icmp->is_propinfo_tbl :\
+	(ns)->netstack_tcp->tcps_propinfo_tbl)
+static int
+lxpr_write_netstacks_property(lxpr_node_t *lxpnp, struct uio *uio,
+    struct cred *cr, caller_context_t *ct, char *prop,
+    int (*xlate)(char *, int))
+{
+	int error;
+	int res = 0;
+	size_t olen, proto_cnt, i;
+	unsigned long newval;
+	uint32_t min, max;
+	char val[16];	/* big enough for a uint numeric string */
+	char *end;
+	netstack_t *ns;
+	mod_prop_info_t *ptbl = NULL;
+	mod_prop_info_t *pinfo = NULL;
+	mod_prop_info_t *maxbuf_pinfo = NULL;
+	uint_t proto_entries[] = {
+		MOD_PROTO_TCP,
+		MOD_PROTO_UDP,
+		MOD_PROTO_SCTP,
+		MOD_PROTO_RAWIP
+	};
+
+	if (uio->uio_loffset != 0)
+		return (EINVAL);
+
+	if (uio->uio_resid == 0)
+		return (0);
+
+	olen = uio->uio_resid;
+	if (olen > sizeof (val) - 1)
+		return (EINVAL);
+
+	bzero(val, sizeof (val));
+
+	error = uiomove(val, olen, UIO_WRITE, uio);
+
+	if (error != 0)
+		return (error);
+
+	if (val[olen - 1] == '\n')
+		val[olen - 1] = '\0';
+
+	if (val[0] == '\0') /* no input */
+		return (EINVAL);
+
+	ns = lxpr_netstack(lxpnp);
+	if (ns == NULL)
+		return (EINVAL);
+
+	if (xlate != NULL && xlate(val, sizeof (val)) != 0) {
+		netstack_rele(ns);
+		return (EINVAL);
+	}
+
+	if (ddi_strtoul(val, &end, 10, &newval) != 0 || *end != '\0')
+		return (EINVAL);
+	/*
+	 * We are trying to catch ERANGE errors when updating the
+	 * property, so we first get the min and max values for
+	 * the netstack and if it's out of range we do bail out.
+	 */
+
+	proto_cnt = sizeof (proto_entries)/ sizeof (proto_entries[0]);
+	for (i = 0; i < proto_cnt; i++) {
+		ptbl = PTBL_FROM_NETSTACK(ns, proto_entries[i]);
+		pinfo = mod_prop_lookup(ptbl, prop, proto_entries[i]);
+		if (pinfo == NULL) {
+			netstack_rele(ns);
+			return (EINVAL);
+		}
+
+		maxbuf_pinfo = mod_prop_lookup(ptbl, "max_buf",
+		    pinfo->mpi_proto);
+		if (maxbuf_pinfo == NULL) {
+			netstack_rele(ns);
+			return (EINVAL);
+		}
+
+		min = pinfo->prop_min_uval;
+		max = maxbuf_pinfo->prop_max_uval;
+
+		if (newval < min || newval > max) {
+			netstack_rele(ns);
+			return (EINVAL);
+		}
+	}
+
+	/*
+	 * At this point the value supplied should be in the possible ranges.
+	 * So we apply the new value to the property specified.
+	 */
+
+	(void) snprintf(val, sizeof (val), "%ld", newval);
+
+	for (i = 0; i < proto_cnt; i++) {
+		ptbl = PTBL_FROM_NETSTACK(ns, proto_entries[i]);
+		pinfo = mod_prop_lookup(ptbl, prop, proto_entries[i]);
+		if (pinfo == NULL) {
+			netstack_rele(ns);
+			return (EINVAL);
+		}
+		if (pinfo->mpi_setf(ns, cr, pinfo, NULL, val, 0) != 0) {
```

Comment:
Is returning EINVAL in order to be consistent with Linux semantics about this?  Or should we be using pinfo->mpi_setf()'s return (which is 0 or an errno) to Do Something with it?

Reply:
I guess EINVAL it is.

ON THE OTHER HAND... YOU ARE MISSING THE FAILURE MODE(s)... if I set some of the entries (TCP, UDP) but fail on one of the later ones (SCTP or IP), I return failure, but I do NOT revert the ones that got successfully set.  I mentioned this earlier, I think, and it's not addressed here.